### ISSUE TITLE
Allow for anonymous websocket services

### ISSUE THREAD

**Issue by Avnerus** [feathersjs/feathers-passport#3]
Hi, I'm not sure if I understand this correctly, but it seems that unless a user is logged on through Passport, he cannot access any feathers services through a websocket.
But why isn't the authorization delegated to the service level?
Like in the example here: http://feathersjs.com/#toc14
It seems that in here https://github.com/jfromaniello/passport.socketio they allow this scenario by delegating the ```fail()``` function (you can then accept the connection anyhow by calling ```accept()```).
 
Thanks
/Avner

**Comment by daffl** [comment]
I would think that you can just return an anonymous user in the authentication callback and then check that in your service:

```js
passport.use(new LocalStrategy(function(username, password, done) {
  done({ anonymous: true });
});

app.use('/todos', {
  get: function(id, params, callback) {
    if(params.user.anonymous) {
      return callback(null, { id: id, text: 'Anonymous Todo' });
    }
  }
});
```

There is also a [passport-anonymous](https://github.com/jaredhanson/passport-anonymous) strategy.

**Comment by Avnerus** [comment]
The thing is that I do want to support regular user login through LocalStrategy.
I just want some services to work even before the user logs in.
I guess in your example I could call ```req.logIn('anonymous')``` on startup but it feels a bit weird.
If the REST services are runnable without having to log in, why wouldn't the socket services be?

**Comment by daffl** [comment]
Socket services should be accessible without logging in. All the code really does is setting `params.user` in a SocketIO middleware based on the authenticated user (https://github.com/feathersjs/feathers-passport/blob/master/lib/passport.js#L75). Do you have an example that doesn't work?

**Comment by Avnerus** [comment]
Even when running the example in this repository the socket never connects unless you login.
The authorizer blocks the connection here: https://github.com/feathersjs/feathers-passport/blob/master/lib/authorizer.js#L41
This generates the ```socket.on('error')``` event and the todo service is never called.
In my example I used Primus and I get blocked in the same way when trying to make a connection.

**Comment by daffl** [comment]
True indeed. It should be easy to just add a flag and pass an anonymous user object (or something similar). I will look into it.

**Comment by Avnerus** [comment]
Thanks! In my fork I just set the user as null as I don't really need an anonymous user and I can check ```params.user``` in the service (https://github.com/ListnPlay/feathers-passport/commit/a937b5deb396cf811ba0920bd755be3a0cd449e5)

**Comment by daffl** [comment]
The way you are doing it in your fork is totally the right way to do it. I will add some more tests and hopefully make another release this week.

**Comment by Avnerus** [comment]
Great! Thank you for that.
There was another change I had to do (https://github.com/ListnPlay/feathers-passport/commit/e135f14e3744657ead9bca7c854d774cbd1b4b43) for a more unique case. I wrote an example for an isomorphic application so I needed the ability for the server to connect to itself with a socket and fetch data (So the same code can run on both the server and the client). Then there was no session at all so I had to remove another ```Error```. Might be worth considering as well if you're working on that segment.
